src/aws-manager.js

```diff
@@ -139,10 +125,172 @@ class AwsManager {
     this.reportAmiUsage = series.amiUsage.reporter(influx);
   }
 
+  async init(){
+    try {
+      let rawJson = await this.spotRequestContainer.read('internal-provisioner-data');
+      let data = JSON.parse(rawJson);
+      // This is mainly for debugging
+      fs.writeFileSync('internal-data.json', data);
+      this.__spotRequestIdCache = data.managedSpotRequests || []; 
+      this.__internalState = data.internallyTrackedSpotRequests || [];
+    } catch (err) {
+      if (err.code !== 'BlobNotFound') {
+        throw err;
+      }
+      this.__spotRequestIdCache = [];
+      this.__internalState = [];
+    }
+  }
+
+  async saveAwsManagerInternalState() {
+    // We use this for both of the current internal caches... let's just use
+    // the same code for both
+    function expire(item) {
+      let diff = Date.now() - item.created;
+      // If we've known about this for more than 5 days, evict it, otherwise
+      // keep it.  5 days is a good amount since we force kill things even if
+      // they're running after 4 days.  This lets us have a full extra day
+      if (diff > 1000 * 60 * 60 * 24 * 5) {
+        return false;
+      }
+      return true;
+    }
+
+    this.__spotRequestIdCache = this.__spotRequestIdCache.filter(expire);
+
+    let data = JSON.stringify({
+      managedSpotRequests: this.__spotRequestIdCache || [],
+      internallyTrackedSpotRequests: this.__internalState || [],
+    }, null, 2);
+
+    fs.writeFileSync('internal-data.json', data);
+    return this.spotRequestContainer.write('internal-provisioner-data', data);
+  }
+
+  /**
+   * Simplify how we access stored spot request ids
+   */
+  async workerTypeForResource(resource, region) {
+    assert(typeof resource === 'object');
+    assert(typeof region === 'string');
+
+    let srid = resource.SpotInstanceRequestId;
+    let instanceId = resource.InstanceId;
+
+    // First, we should check to see if we have the mapping of
+    // SpotInstanceRequestId in memory (which we persist to an azure blob
+    // storage blob).  If this item is in that cache, we can be sure that the
+    // provisioner both owns it because it created it, and also which worker
+    // type, since we add that information to the cache when we add the id.
+    let found = this.__spotRequestIdCache.filter(x => x.id === srid && x.region === region);
+    if (found.length === 1) {
+      return found[0].workerType;
+    } else if (found.length > 1) {
+      let firstX = found[0].workerType;
```

Comment:
I think that's reasonable.  I just don't want to cause the provisioner to barf if an identical item is in there twice.  I'll look at this shortly.